### PR TITLE
Try latest java.jj and fix some TEXT_BLOCK issues

### PR DESCRIPTION
**Pull request by matozoid**

### CODE REVIEW

File: src/main/antlr/JavaLexer.tokens [hunk 1]
```diff
@@ -42,208 +42,95 @@ PACKAGE=41
...
-'byte'=12
-'case'=13
-'catch'=14
-'char'=15
-'class'=16
-'const'=17
-'continue'=18
-'default'=19
-'do'=20
-'double'=21
```
Comment (matozoid) [inline src/main/antlr/JavaLexer.tokens:149]: Looks like @MysterAitch cleaned up the grammar quite well!

File: src/main/kotlin/com/strumenta/javacc/AntlrModel.kt [hunk 2]
```diff
@@ -10,6 +10,7 @@ data class RuleDefinition(val name: String, val body: String, val action: String
     fun generate() : String {
         val prefix = if (fragment) "fragment " else ""
         val actionPostfix = if (action == null) "" else "-> $action"
+        val body= if(action!=null && body.contains("|")) "($body)" else body
```
Comment (matozoid) [inline src/main/kotlin/com/strumenta/javacc/AntlrModel.kt:13]: Putting () around the body only happens for text blocks.

File: src/main/kotlin/com/strumenta/javacc/AntlrModel.kt [hunk 3]
```diff
@@ -66,8 +70,8 @@ class LexerDefinitions(val name: String) {
         if (ruleDefinitionCorrected.name == ruleDefinitionCorrected.body) {
             return
         }
-        if (ruleDefinitionCorrected.body == "~[]") {
-            ruleDefinitionCorrected = ruleDefinitionCorrected.copy(body = ".")
+        if (ruleDefinitionCorrected.body.contains("~[]")) {
```
Comment (matozoid) [inline src/main/kotlin/com/strumenta/javacc/AntlrModel.kt:70]: TEXT_BLOCK has a more complex body, but it does contain ~[]

File: src/main/kotlin/com/strumenta/javacc/JavaCCToAntlrConverter.kt [hunk 4]
```diff
@@ -133,7 +133,8 @@ private fun generateLexerDefinitions(name: String, tokenDefinitions: List<TokenP
             "MORE" -> {
                 it.respecs.forEach {
                     if (it.nextState != null) {
-                        lexStates.forEach { ls -> lexerDefinitions.addRuleDefinition(ls, it.toRuleDefinition(ls, "pushMode(${it.nextState})")) }
+                        val action = if (it.nextState == DEFAULT_MODE_NAME) "popMode" else  "pushMode(${it.nextState})"
+                        lexStates.forEach { ls -> lexerDefinitions.addRuleDefinition(ls, it.toRuleDefinition(ls, action)) }
```
Comment (matozoid) [inline src/main/kotlin/com/strumenta/javacc/JavaCCToAntlrConverter.kt:137]: TEXT_BLOCK is used by the parser, it's not SPECIAL.

File: src/main/kotlin/com/strumenta/javacc/AntlrModel.kt [hunk 5]
```diff
@@ -91,9 +95,10 @@ class LexerDefinitions(val name: String) {
             printWriter.println("mode $mode;")
         }
         rulesByMode[mode]!!.forEach {
-            if (it.name.contains("COMMENT") && it.action == null) {
+            val specialHandlingRequired = it.name.contains("COMMENT") || it.name.contains("TEXT_BLOCK")
+            if (specialHandlingRequired && it.action == null) {
                 printWriter.println(it.copy(action = "skip").generate())
-            } else if (it.name.contains("COMMENT") && it.action != null && !it.action.contains("skip")) {
+            } else if (specialHandlingRequired && it.action != null && !it.action.contains("skip")) {
```
Comment (matozoid) [inline src/main/kotlin/com/strumenta/javacc/AntlrModel.kt:101]: I think this is wrong...

**Review by ftomassetti — APPROVED** [review]

**Comment by ftomassetti** [comment]
Thank you for the PR!

**Comment by matozoid** [comment]
Note that this doesn't completely resolve the problems with TEXT_BLOCKs. The Antlr code delivers different tokens than JavaCC. That was already occurring in the original, and the PR made it less bad, but the new java.jj triggers the problem. It's easy to work around though. I was going to fix the tests, but this will need a good amount of work that I suspect nobody is willing to put in ;-)

**Pull request merged**